File: Backtracking/17.letter-combinations-of-a-phone-number.py
#
# @lc app=leetcode id=17 lang=python3
#
# [17] Letter Combinations of a Phone Number
#

# @lc code=start
class Solution:
    def __init__(self):
        self.letterMap = [
            "",
            "",
            "abc",
            "def",
            "ghi",
            "jkl",
            "mno",
            "pqrs",
            "tuv",
            "wxyz"
        ]

    def letterCombinations(self, digits: str) -> List[str]:
        def backtracking(index, combo):
            if index == len(digits):
                res.append(combo[:])
                return

            digit = int(digits[index])
            letters = self.letterMap[digit]

            # for i in range(len(letters)):
            #     combo += letters[i]
            #     backtracking(index + 1, combo)
            #     combo = combo[:-1]

            for letter in letters:
                combo += letter
                backtracking(index + 1, combo)
                combo = combo[:-1]


        res = []
        if len(digits) == 0:
            return res
        backtracking(0, "")
        return res
# @lc code=end

File: Backtracking/17.letter-combinations-of-a-phone-number.java

/*
 * @lc app=leetcode id=17 lang=java
 *
 * [17] Letter Combinations of a Phone Number
 */
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

// @lc code=start
class Solution {
    private String[] numMap = {
            "",
            "",
            "abc",
            "def",
            "ghi",
            "jkl",
            "mno",
            "pqrs",
            "tuv",
            "wxyz"
    };

    private List<String> res = new ArrayList<>();

    private void recHelper(String digits, int index, String curStr) {
        if (index == digits.length()) {
            res.add(curStr);
            return;
        }
        String letters = numMap[digits.charAt(index) - '0'];
        for (int i = 0; i < letters.length(); i++) {
            recHelper(digits, index + 1, curStr + letters.charAt(i));
        }
    }

    public List<String> letterCombinations(String digits) {
        if (digits.length() == 0) {
            return res;
        }
        recHelper(digits, 0, "");
        return res;
    }
}
// @lc code=end
